### DISCUSSION TITLE
CodeQL Queries in LGTM

### DISCUSSION THREAD

**Discussion by ArielSAdamsNASA** [community/community#24711] [Apps, API and Webhooks]
<p>I am wanting to place JPL queries from CodeQL to LGTM.</p>
<p>What queries does LGTM have? Where do I find a list of these queries?</p>
<p>Is there a way to use certain CodeQL queries in LGTM besides downloading the queries and placing them in a .lgtm folder as custom queries?</p>
<p>Can LGTM take in a SARIF file and upload its results?</p>

**Comment by MathiasVP** [comment, marked as answer]
<p>Hi ArielSAdamsNASA,</p>
<blockquote>
<p>What queries does LGTM have? Where do I find a list of these queries?</p>
</blockquote>
<p>By default, LGTM will run all queries that satisfy the following criteria:</p>
<ul>
<li>The <code>@problem.severity</code> is <code>Error</code> or <code>Warning</code>, and the <code>@precision</code> is at least <code>Medium</code>
</li>
<li>The <code>@problem.severity</code> is <code>Recommendation</code> and the <code>@precision</code> is at least <code>High</code>
</li>
</ul>
<p>And the queries that satisfy the following criteria are being displayed:</p>
<ul>
<li>The <code>@problem.severity</code> is <code>Error</code> or <code>Warning</code>, and the <code>@precision</code> is at least <code>High</code>
</li>
<li>The <code>@problem.severity</code> is <code>Recommendation</code> and the <code>@precision</code> is at least <code>Very high</code>
</li>
</ul>
<p>A quick-and-dirty way of getting a list of the queries on LGTM is to search for queries with an <code>@id</code> that starts with <code>cpp/</code> (to get a list of all the C++ queries). You can see this working <a href="https://lgtm.com/search?q=cpp%2F&amp;t=rules">here</a>.</p>
<blockquote>
<p>Is there a way to use certain CodeQL queries in LGTM besides downloading the queries and placing them in a .lgtm folder as custom queries?</p>
</blockquote>
<p>You can configure which queries should be run by using a custom lgtm.yml configuration file. You can download a comprehensive template lgtm.yml file from <a href="https://lgtm.com/help/lgtm/lgtm.yml-configuration-file">here</a> that explains this. Specifically, you should look at the <code>queries</code> block:</p>
<pre><code class="lang-yaml">queries:
  # Start by hiding the results of all queries.
  - exclude: "*"
  # Then include all queries tagged 'security' and 'correctness', and with a severity of
  # 'error'.
  - include:           
      tags:
        - "security"
        - "correctness"           
      severity: "error"
  # Specifically hide the results of two queries.
  - exclude: cpp/use-of-goto
  - exclude: java/equals-on-unrelated-types
  # Refine by including the `java/command-line-injection` query.
  - include: java/command-line-injection
</code></pre>
<blockquote>
<p>Can LGTM take in a SARIF file and upload its results?</p>
</blockquote>
<p>Unfortunately, no. You can’t upload a SARIF file and display the results in LGTM.<br>
It is, however, possible to do this with Code Scanning.</p>
<p>I hope these answers are helpful. If not, I’ll be happy to expand on any of them!</p>

**Comment by ArielSAdamsNASA** [comment]
<p>Would it be possible to create the lgtm/cpp-queries in .lgtm.yml like so instead of creating a permanent folder in github?</p>
<pre><code class="lang-auto">queries:
- include: "*"
- exclude: "cpp/class-many-fields"
- exclude: "cpp/long-switch"
- exclude: "cpp/trivial-switch"

path_classifiers:
  generated:
    - "**/Ui_*"

extraction:
  python:
    python_setup:
      version: "3"
    index:
      include: "tools/cFS-GroundSystem"
  cpp:
    index:
      build_command:
      - "git clone https://github.com/github/codeql.git codeql"
      - "mkdir -p lgtm/cpp-queries/
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-2/Rule 03/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-2/Rule 04/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-2/Rule 05/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-2/Rule 07/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-2/Rule 09/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-2/Rule 11/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-3/Rule 12/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-3/Rule 13/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-3/Rule 14/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-3/Rule 15/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-3/Rule 16/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-3/Rule 17/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-3/Rule 18/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-3/Rule 20/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 21/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 22/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 23/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 24/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 25/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 26/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 27/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 28/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 29/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 30/.' lgtm/cpp-queries"
      - "cp -a 'codeql/cpp/ql/src/JPL_C/LOC-4/Rule 31/.' lgtm/cpp-queries"
      - "cp cfe/cmake/Makefile.sample Makefile"
      - "cp -r cfe/cmake/sample_defs sample_defs"
      - "make OMIT_DEPRECATED=true ENABLE_UNIT_TEST=true prep"
      - "make"

</code></pre>

**Comment by rneatherway** [comment]
<p>I’m afraid that this is not possible. LGTM looks for these custom queries when polling the repository for new commits. This is mentioned in passing at <a href="https://lgtm.com/help/lgtm/about-queries#how-do-i-add-custom-queries-to-LGTM" class="inline-onebox" rel="noopener nofollow ugc">LGTM</a></p>
<blockquote>
<p>The daily LGTM poll jobs will detect the queries and automatically run them.</p>
</blockquote>

**Comment by rneatherway** [comment]
<p>I should add that this <em>is</em> possible using Code Scanning: <a href="https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#running-additional-queries" class="inline-onebox" rel="noopener nofollow ugc">Configuring code scanning - GitHub Docs</a>.</p>
<p>CodeQL is normally run using Actions, in which case it is easy to arrange the order of operations such that you put the queries in place before performing the analysis. Alternatively if you use another CI system it is typical to invoke the CodeQL CLI yourself and the same applies.</p>
